src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+    Posix,
+    Bash,
+    Zsh,
+}
+
+pub fn get_available_shells() -> impl Iterator<Item = Shell> {
+    enumerate_shells().into_iter().filter(|sh| sh.does_exist())
+}
+
+pub trait UnixShell {
+    // Detects if a shell "exists". Users have multiple shells, so an "eager"
+    // heuristic should be used, assuming shells exist if any traces do.
+    fn does_exist(&self) -> bool;
+
+    // Gives all rcfiles of a given shell that rustup is concerned with.
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
+
+    // Gives rcs that should be written to.
+    fn update_rcs(&self) -> Vec<PathBuf>;
+
+    // Writes the relevant env file.
+    fn env_script(&self) -> ShellScript {
+        ShellScript {
+            name: "env",
+            content: POSIX_ENV,
+        }
+    }
+
+    fn source_string(&self) -> Result<String> {
+        Ok(format!(r#"source "{}/env""#, canonical_cargo_home()?))
+    }
+}
+
+struct Posix;
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match utils::home_dir() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist.
+        self.rcfiles()
+    }
+}
+
+struct Bash;
+
+// Bash is the source of many complications because it loads ONLY 1 rc,
+// either a login profile or .bashrc, not both.
+impl Bash {
+    // Bash will load only one of these, the first one that exists, when
+    // a login shell starts. BUT not all shells start inside a login shell!
+    fn profiles() -> Vec<PathBuf> {
+        [".bash_profile", ".bash_login", ".profile"]
+            .iter()
+            .filter_map(|rc| utils::home_dir().map(|dir| dir.join(rc)))
+            .collect()
+    }
+
+    // Bash will load only .bashrc on the start of most GUI terminals.
+    fn rc() -> Option<PathBuf> {
+        utils::home_dir().map(|dir| dir.join(".bashrc"))
+    }
+}
+
+impl UnixShell for Bash {
+    fn does_exist(&self) -> bool {
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("bash"))
+            || self.rcfiles().iter().any(|rc| rc.is_file())
+            || matches!(utils::find_cmd(&["bash"]), Some(_))
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        let mut profiles = Bash::profiles();
+        if let Some(rc) = Bash::rc() {
+            profiles.push(rc);
+        }
+        profiles
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        Bash::profiles()
+            .into_iter()
+            .filter(|rc| rc.is_file())
+            // bash only reads one "login profile" so pick the one that exists
+            .take(1)
+            // Pick .bashrc if it exists for GUI terminals
+            .chain(Bash::rc().filter(|rc| rc.is_file()))
+            .collect()
+    }
+}
+
+struct Zsh;
+impl UnixShell for Zsh {
+    fn does_exist(&self) -> bool {
```

Comment:
I'm not completely clear on what this actually tests; I *think* it is:
if zsh in SHELL or ZDOTDIR != "" or ~/.zshenv exists or zsh exists.

Reply:
Yes.